### ISSUE TITLE
Add a warning about Numba and Python 3.5 compatibility.

### ISSUE THREAD

**Issue by tamasgal** [scikit-hep/awkward#219]
The problem I encountered is that `numba.core.typing` is not present e.g. in numba v0.47 which is the latest for Python 3.5.

I found the corresponding commit: https://github.com/scikit-hep/awkward-1.0/commit/3df1e14622cdc1f7561914252f19f5b39ffde2ae

This was probably not caught by the awkward1 test suite. Here is the error which came up in my library:

```python
venv/lib/python3.5/site-packages/awkward1/_connect/_numba/__init__.py:10: in register
    import awkward1._connect._numba.arrayview
venv/lib/python3.5/site-packages/awkward1/_connect/_numba/arrayview.py:9: in <module>
    import numba.core.typing
E   ImportError: No module named 'numba.core'
```

I thought making a PR which replaces `import numba.core.typing` with `import numba.typing` etc. (in `_connect/_numba/arrayview.py` and `_connect/_numba/builder.py`) would solve the problem but apparently `numba.typing.ctypes` does not exist in 0.47, which is also used:

```python
In [1]: import numba.typing

In [2]: numba.typing.ctypes_utils
---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
<ipython-input-2-d7a62042b95a> in <module>
----> 1 numba.typing.ctypes_utils

AttributeError: module 'numba.typing' has no attribute 'ctypes_utils'

In [3]: numba.__version__
Out[3]: '0.47.0'
```

What works is:

```python
In [1]: from numba.typing import ctypes_utils
```

This on the other hand will not work for numba 0.50+

```python
In [1]: from numba.typing import ctypes_utils
/usr/local/bin/ipython:1: NumbaDeprecationWarning: An import was requested from a module that has moved location.
Import of 'ctypes_utils' requested from: 'numba.typing', please update to use 'numba.core.typing' or pin to Numba version 0.48.0. This alias will not be present in Numba version 0.50.0.
  #!/usr/local/bin/python
```

What do you think about some special code for Python 3.5 (`ImportError` catches or version check blocks and alike)?

**Comment by jpivarski** [comment]
That's true, and I punted on supporting Numba with Python 3.5. Since Numba is not an explicit dependency of Awkward, I couldn't put it in the requirements.txt, but the "dev requirements" includes

```
numba>=0.49.0;python_version>="3.6
```

What you're seeing is that

   * Numba moved the locations of most of the classes and functions useful for extensions like ours to present a more stable public API. This was, in a large part, inspired by a [talk I gave to the Numba group](https://nbviewer.jupyter.org/github/scikit-hep/awkward-1.0/blob/master/docs-demo-notebooks/2020-01-22-numba-demo-EVALUATED.ipynb#4.2-Survey-of-Numba-functions-used) which led to them [conducting a survey](https://github.com/numba/numba/issues/5214) of all the major users of Numba extensions to figure out which are the important functions to present in an API. Since everything moved, our minimum Numba version is now 0.49, after the move.
   * The last NumPy version that can run on Python 3.5 is 1.14 (see [NEP 29](https://numpy.org/neps/nep-0029-deprecation_policy.html)).
   * The earliest NumPy version that Numba 0.49 can use is 1.15 (see [release notes](https://numba.pydata.org/numba-doc/dev/release-notes.html#version-0-49-0-apr-16-2020)).

Hence, Numba won't work in Awkward for Python 3.5. QED.

I ought to put some specific check in, to provide a clearer warning to the user.

**Comment by jpivarski** [comment]
I changed the name of this issue because what I plan to do with this is to make Awkward complain on entry (the `awkward1._connect._numba.register` function) if the Numba version is less than 0.49, which would always be true for Python 3.5.

It won't complain on import (so that the mere existence of an old Numba doesn't break Awkward), but when you attempt to use an array in a Numba function.

**Comment by tamasgal** [comment]
Alright, yes I was aware of these changes and the dependency issues. Just wanted to let you know that this bubbles up in an inconvenient way, but thanks for the summary!

> I ought to put some specific check in, to provide a clearer warning to the user.

OK that sounds fine!

**Comment by jpivarski** [comment]
Fixed in PR #227.

**Comment by tamasgal** [comment]
Thanks Jim!
